### PR TITLE
 Introduced flag to specify when sts deleted, pvc's behavious

### LINKED ISSUE #4750: fix:  kubectl  karmada deinit  doesn't cascade deletion of pvc resources
**What happened**:
1. Create karmada control plane as as the following command:
`kubectl-karmada init --kube-image-registry=registry.cn-hangzhou.aliyuncs.com/google_containers --storage-classes-name=local-path  --etcd-pvc-size='30Gi' --etcd-storage-mode='PVC'  --wait-component-ready-timeout=360 --etcd-replicas 3   --karmada-apiserver-replicas 3`
2. pod and pvc created successfully
image
3. remove karmada control plane as  the following command:
`kubectl  karmada deinit`
4. pod deleted but pvc not
image

**What you expected to happen**:

pvc should  be cascade deleted when etcd pods deleted.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Karmada version:   karmada 1.9
- kubectl-karmada or karmadactl version (the result of `kubectl-karmada version` or `karmadactl version`):
- Others:

### PR DESCRIPTION
**Pull request by CharlesQQ**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
 Introduced flag to specify when sts deleted, pvc's behavious

**Which issue(s) this PR fixes**:
Fixes https://github.com/karmada-io/karmada/issues/4750

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

`karmadactl`/`kubectl-karmada`: Introduced `etcd-sts-pvc-delete-policy` flags to specify when sts deleted, pvc's behavious
```

### CODE REVIEW

File: pkg/karmadactl/cmdinit/cmdinit.go [hunk 1]
```diff
@@ -131,13 +131,14 @@ func NewCmdInit(parentCommand string) *cobra.Command {
...
+		fmt.Sprintf("etcd data storage mode(%s). value is PVC, specify --storage-classes-name", strings.Join(kubernetes.SupportedStorageMode().List(), ",")))
 	flags.StringVarP(&opts.EtcdImage, "etcd-image", "", "", "etcd image")
 	flags.StringVarP(&opts.EtcdInitImage, "etcd-init-image", "", kubernetes.DefaultInitImage, "etcd init container image")
 	flags.Int32VarP(&opts.EtcdReplicas, "etcd-replicas", "", 1, "etcd replica set, cluster 3,5...singular")
 	flags.StringVarP(&opts.EtcdHostDataPath, "etcd-data", "", "/var/lib/karmada-etcd", "etcd data path,valid in hostPath mode.")
 	flags.StringVarP(&opts.EtcdNodeSelectorLabels, "etcd-node-selector-labels", "", "", "etcd pod select the labels of the node. valid in hostPath mode ( e.g. --etcd-node-selector-labels karmada.io/etcd=true)")
 	flags.StringVarP(&opts.EtcdPersistentVolumeSize, "etcd-pvc-size", "", "5Gi", "etcd data path,valid in pvc mode.")
+	flags.StringVarP(&opts.EtcdStsPvcDeletePolicy, "etcd-sts-pvc-delete-policy", "", "Retain", fmt.Sprintf("(%s). retain or delete pvc when their stateful set is deleted. Retain, need k8s >= v1.27", strings.Join(kubernetes.SupportedPvcRetentionPolicyType().List(), ",")))
```
Comment (XiShanYongYe-Chang) [inline pkg/karmadactl/cmdinit/cmdinit.go:141]: Hi @chaosi-zju @prodanlabs @liangyuanpeng @RainbowMango How do you think this new flag?

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go [hunk 2]
```diff
@@ -733,6 +735,11 @@ func generateServerURL(serverIP string, nodePort int32) (string, error) {
 }
 
 // SupportedStorageMode Return install etcd supported storage mode
-func SupportedStorageMode() []string {
-	return []string{etcdStorageModeEmptyDir, etcdStorageModeHostPath, etcdStorageModePVC}
+func SupportedStorageMode() sets.String {
```
Comment (XiShanYongYe-Chang) [inline pkg/karmadactl/cmdinit/kubernetes/deploy.go:738]: set.String is deprecated, how about use `sets.Set[string]` to replace it?

File: pkg/karmadactl/cmdinit/kubernetes/statefulset.go [hunk 3]
```diff
@@ -351,14 +351,22 @@ func (i *CommandInitOption) makeETCDStatefulSet() *appsv1.StatefulSet {
 		Spec: podSpec,
 	}
 
+	var stsPvcRetentionPolicy *appsv1.StatefulSetPersistentVolumeClaimRetentionPolicy
+	if i.EtcdStsPvcDeletePolicy == string(appsv1.DeletePersistentVolumeClaimRetentionPolicyType) {
+		stsPvcRetentionPolicy = &appsv1.StatefulSetPersistentVolumeClaimRetentionPolicy{
+			WhenDeleted: appsv1.DeletePersistentVolumeClaimRetentionPolicyType,
+			WhenScaled:  appsv1.RetainPersistentVolumeClaimRetentionPolicyType,
```
Comment (XiShanYongYe-Chang) [inline pkg/karmadactl/cmdinit/kubernetes/statefulset.go:358]: I have a question, I don't understand why this is set up.

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go (hunk 2 above)
Comment (CharlesQQ) [reply pkg/karmadactl/cmdinit/kubernetes/deploy.go:738]: > how about use sets.Set[string] to replace it?

ok

File: pkg/karmadactl/cmdinit/kubernetes/statefulset.go (hunk 3 above)
Comment (CharlesQQ) [reply pkg/karmadactl/cmdinit/kubernetes/statefulset.go:358]: > why this is set up.

Default policy is appsv1.RetainPersistentVolumeClaimRetentionPolicyType; if set flag, WhenDeleted value is appsv1.DeletePersistentVolumeClaimRetentionPolicyType

File: pkg/karmadactl/cmdinit/cmdinit.go [hunk 4]
```diff
@@ -131,13 +131,14 @@ func NewCmdInit(parentCommand string) *cobra.Command {
...
+		fmt.Sprintf("etcd data storage mode(%s). value is PVC, specify --storage-classes-name", strings.Join(kubernetes.SupportedStorageMode().UnsortedList(), ",")))
 	flags.StringVarP(&opts.EtcdImage, "etcd-image", "", "", "etcd image")
 	flags.StringVarP(&opts.EtcdInitImage, "etcd-init-image", "", kubernetes.DefaultInitImage, "etcd init container image")
 	flags.Int32VarP(&opts.EtcdReplicas, "etcd-replicas", "", 1, "etcd replica set, cluster 3,5...singular")
 	flags.StringVarP(&opts.EtcdHostDataPath, "etcd-data", "", "/var/lib/karmada-etcd", "etcd data path,valid in hostPath mode.")
 	flags.StringVarP(&opts.EtcdNodeSelectorLabels, "etcd-node-selector-labels", "", "", "etcd pod select the labels of the node. valid in hostPath mode ( e.g. --etcd-node-selector-labels karmada.io/etcd=true)")
 	flags.StringVarP(&opts.EtcdPersistentVolumeSize, "etcd-pvc-size", "", "5Gi", "etcd data path,valid in pvc mode.")
+	flags.StringVarP(&opts.EtcdStsPvcDeletePolicy, "etcd-sts-pvc-delete-policy", "", "Retain", fmt.Sprintf("(%s). retain or delete pvc when their stateful set is deleted. Retain, need k8s >= v1.27", strings.Join(kubernetes.SupportedPvcRetentionPolicyType().UnsortedList(), ",")))
```
Comment (XiShanYongYe-Chang) [inline pkg/karmadactl/cmdinit/cmdinit.go:141]: Can we directly provide two flags to set the `WhenDeleted` and `WhenScaled` rotation policy type respectively without processing an extra layer and changing the original meaning?

**Comment by XiShanYongYe-Chang** [comment]
What should we do with these flags for  Kubernetes versions prior to v1.26?

**Comment by CharlesQQ** [comment]
> What should we do with these flags for Kubernetes versions prior to v1.26?

patch ovnerreference to pvc metedata

**Comment by XiShanYongYe-Chang** [comment]
> patch ovnerreference to pvc metedata

Sorry, let me rephrase my question. If the version of karmada-apiserver is before 1.27, will the current change have no effect or will it cause the etcd statefulset to report an error?

**Comment by CharlesQQ** [comment]
>  If the version of karmada-apiserver is before 1.27, will the current change have no effect or will it cause the etcd statefulset to report an error?

sts has no field spec.persistentVolumeClaimRetentionPolicy, so sts can't be created and report an error

**Comment by XiShanYongYe-Chang** [comment]
This behavior may not be expected as we allow the user to set the version of the apiserver:

https://github.com/karmada-io/karmada/blob/456d099404d09ba4094614a6a3a8bea36cdf6d61/pkg/karmadactl/cmdinit/cmdinit.go#L121

We may need to add some judgment logic.
